### PR TITLE
tests: New test for xtimer_now with interrupts disabled

### PR DESCRIPTION
**Pull request by JulianHolzwarth**
### Contribution description
This is a new test for xtimer. It tests if xtimer now (`xtimer_now_usec`) returns the correct value when interrupts are disabled.

This test is for: #9530 

### Testing procedure
`BOARD=* make -C tests/xtimer_now_irq/ flash term` replace * with board
If it returns an Error it is not correct.

### Issues/PRs references

#9530

### CODE REVIEW

File: tests/xtimer_now_irq/Makefile.ci [hunk 1]
```diff
@@ -0,0 +1,10 @@
+BOARD_INSUFFICIENT_MEMORY := \
```
Comment (MichelRottleuthner) [inline tests/xtimer_now_irq/Makefile.ci]: we should re-check this because the test is actually rather small

File: tests/xtimer_now_irq/main.c [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+#include <stdio.h>
+#include "xtimer.h"
+#include "irq.h"
+
+int main(void)
+{
+    puts("xtimer_now_irq test application.\n");
+
+    while (true) {
```
Comment (aabadie) [inline tests/xtimer_now_irq/main.c]: Instead of an infinite loop, use a `count--` with count initialized with a sensible value.
Then print "SUCCESS" when getting out of the while loop and no error happened.

File: tests/xtimer_now_irq/main.c [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        unsigned int state = irq_disable();
+        uint32_t t1 = xtimer_now_usec();
+        xtimer_spin(xtimer_ticks_from_usec(XTIMER_BACKOFF));
+        uint32_t t2 = xtimer_now_usec();
+        irq_restore(state);
+        if (t2 < t1) {
+            printf("Error\n");
```
Comment (aabadie) [inline tests/xtimer_now_irq/main.c]: Add a `return -1;` after printing the error message. Replace the printf with `puts("ERROR!");`.

File: tests/xtimer_now_irq/main.c [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     tests
+ * @{
+ *
+ * @file
+ * @brief       testing xtimer_mutex_lock_timeout function
```
Comment (aabadie) [inline tests/xtimer_now_irq/main.c]: copy paste from another test

**Comment by fjmolinas** [comment]
ping @aabadie!

File: tests/xtimer_now_irq/tests/01-run.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+# @author      Julian Holzwarth <julian.holzwarth@fu-berlin.de>
+
+import sys
+import pexpect
+from testrunner import run
+
+
+def testfunc(child):
+    child.expect("OK", timeout=75*60)
```
Comment (aabadie) [inline tests/xtimer_now_irq/tests/01-run.py]: The first 4 expects could be called in a for loop and replaced by `expect_exact` since no regexp is used:
```python
for _ in range(4):
    child.expect_exact("OK", timeout=TIMEOUT)
```

Define TIMEOUT outside of the `testfunc` function. By the way, how do determine the timeout value ?

File: tests/xtimer_now_irq/main.c [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+int main(void)
+{
+    test_utils_interactive_sync();
+    puts("xtimer_now_irq test application.\n");
+
+    while (TEST_COUNT) {
+        unsigned int state = irq_disable();
+        uint32_t t1 = xtimer_now_usec();
+        xtimer_spin(xtimer_ticks_from_usec((uint32_t)(~XTIMER_MASK) / 2));
```
Comment (aabadie) [inline tests/xtimer_now_irq/main.c]: Why do you use `XTIMER_MASK` ? Wouldn't it be simpler to directly use a usec value ? This way it would be defined globally and an approaching value could be used in the Python test script.

File: tests/xtimer_now_irq/main.c [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+#include "test_utils/interactive_sync.h"
+
+#define TEST_COUNT 4
+
+int main(void)
+{
+    test_utils_interactive_sync();
+    puts("xtimer_now_irq test application.\n");
+
+    while (TEST_COUNT) {
```
[the remark this answers is not in the document]
Comment (MichelRottleuthner) [reply tests/xtimer_now_irq/main.c]: @JulianHolzwarth  when halve a timer period is used as spin a small number to cover a handful of overflows should already be enough.

File: tests/xtimer_now_irq/tests/01-run.py (hunk 5 above)
Comment (MichelRottleuthner) [reply tests/xtimer_now_irq/tests/01-run.py]: The number is selected to roughly cover a 32 bit timer running at 1 MHz (~72 minutes), right?

File: tests/xtimer_now_irq/main.c (hunk 6 above)
Comment (MichelRottleuthner) [reply tests/xtimer_now_irq/main.c]: Actually the value handed to spin should be ticks, so interpreting the XTIMER_MASK as usecs doesn't make sense. But using XTIMER_MASK so the value is relative to the maximum value of the timer is good to select the value so the overflow can actually be assumed to happen while interrupts are disabled.

File: tests/xtimer_now_irq/tests/01-run.py (hunk 5 above)
Comment (JulianHolzwarth) [reply tests/xtimer_now_irq/tests/01-run.py]: yes @MichelRottleuthner

**Comment by JulianHolzwarth** [comment]
@aabadie @MichelRottleuthner 
Thank you for your comments.
A pushed a new commit. What do you think now?

File: tests/xtimer_now_irq/README.md [hunk 8]
```diff
@@ -0,0 +1,2 @@
+This test will print an error when the time is incorrect while interrups are disabled.
```
Comment (MichelRottleuthner) [inline tests/xtimer_now_irq/README.md]: Please change to something more specific like: This test will print an error when xtimer_now returns an invalid value when called while interrupts are disabled.

File: tests/xtimer_now_irq/README.md [hunk 9]
```diff
@@ -0,0 +1,2 @@
...
+This test works only reasonaly well on more than 16 Bit platforms.
```
Comment (MichelRottleuthner) [inline tests/xtimer_now_irq/README.md]: How bout something like: "This test must run at least over a full timer period (i.e. >72 minutes with a 32 bit timer running at 1 MHz). So it only finishes in reasonable short time on platforms with small timers like 8 bit or 16 bit."

**Comment by JulianHolzwarth** [comment]
@MichelRottleuthner I changed the README last week. What do you think now?

File: tests/xtimer_now_irq/main.c (hunk 6 above)
Comment (JulianHolzwarth) [reply tests/xtimer_now_irq/main.c]: This should be addressed with the new commit. Correct?

Comment (JulianHolzwarth) [reply tests/xtimer_now_irq/main.c]: @MichelRottleuthner

File: tests/xtimer_now_irq/Makefile.ci (hunk 1 above)
Comment (MichelRottleuthner) [reply tests/xtimer_now_irq/Makefile.ci]: I checked the list and it seems to fit all these boards just fine - so why are they listed as insufficient memory?

**Comment by MichelRottleuthner** [comment]
Apart from the `INSUFFICIENT_MEMORY` list in Makefile.ci I think this is good to go. @aabadie are all your requests addressed?

File: tests/xtimer_now_irq/tests/01-run.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+# This file is subject to the terms and conditions of the GNU Lesser
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+
+# @author      Julian Holzwarth <julian.holzwarth@fu-berlin.de>
+
+import sys
+from testrunner import run
+
+TIMEOUT = 75*60
```
Comment (aabadie) [inline tests/xtimer_now_irq/tests/01-run.py]:
```suggestion
TIMEOUT = 75 * 60
```

File: tests/xtimer_now_irq/Makefile.ci (hunk 1 above)
Comment (JulianHolzwarth) [reply tests/xtimer_now_irq/Makefile.ci]: I will remove it.

File: tests/xtimer_now_irq/tests/01-run.py (hunk 10 above)
Comment (JulianHolzwarth) [reply tests/xtimer_now_irq/tests/01-run.py]: Changed it.

**Comment by MichelRottleuthner** [comment]
@aabadie how bout now?

**Comment by kaspar030** [comment]
The pinetime errors  are unrelated, fix for those is here: https://github.com/RIOT-OS/RIOT/pull/13481

**Comment by kaspar030** [comment]
This also needs some squashing.

File: tests/xtimer_now_irq/main.c [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+int main(void)
+{
+    puts("xtimer_now_irq test application.\n");
+
+    uint8_t count = TEST_COUNT;
+    while (count) {
+        unsigned int state = irq_disable();
+        uint32_t t1 = xtimer_now_usec();
+        xtimer_spin(xtimer_ticks((uint32_t)(~XTIMER_MASK) / 2));
```
Comment (kaspar030) [inline tests/xtimer_now_irq/main.c]: On 32bit 1mhz, this will spin `(0xffffffff / 2)` ticks. Doesn't that make the second iteration always fail?

1. xtimer_now() is somewhere a little higher than zero, so is t1
2. spin half period
3. t2 is now somewhere a little higher than a half period
4. next loop iteration, t1 is now close to t2 (a little higher than a half period)
5. spin half period
6. t2 overflows ("a little higher than a half period" + "half period" >> full period)
7. t2 < t1, error triggers

What am I missing?

Comment (MichelRottleuthner) [reply tests/xtimer_now_irq/main.c]: Running it as is for 32 bit timer is madness anyway and we need a better way for testing these corner cases with a bit of mocks magic.
Catching the "missed overflow" of the 32 bit timer with the 32 bit xtimer interface obviously doesn't make sense, the 64 bit variant is needed for that ;)

Comment (kaspar030) [reply tests/xtimer_now_irq/main.c]: Alright, how about straight passing if `XTIMER_WIDTH == 32`?

File: tests/xtimer_now_irq/README.md [hunk 12]
```diff
@@ -0,0 +1,7 @@
+# README for xtimer_now_irq test
+
+This test checks, if the timer returns the correct time (xtimer_now_usec() is called), when interrupts are disabled. Specifically tested is if the time is correct after a low-level timer overflow.
+
+When the returned value is incorrect the test will print: "ERROR: wrong time with interrupts disabled".
+
+This test must be run over at least one full timer period (i.e. >72 minutes with a 32 bit timer at 1MHz). Meaning the test only finishes in reasonable short time on platforms with small timers like 8 bit or 16 bit.
```
Comment (kaspar030) [inline tests/xtimer_now_irq/README.md]: Makes no sense to run this on 32bit/1MHz. Better rephrase.

File: tests/xtimer_now_irq/tests/01-run.py [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+# This file is subject to the terms and conditions of the GNU Lesser
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+
+# @author      Julian Holzwarth <julian.holzwarth@fu-berlin.de>
+
+import sys
+from testrunner import run
+
+TIMEOUT = 75 * 60
```
Comment (kaspar030) [inline tests/xtimer_now_irq/tests/01-run.py]: On non-32bit timers (next known size ios 24bit), shit should fail after 2**24 -> ~16 seconds, right?

File: tests/xtimer_now_irq/main.c (hunk 11 above)
Comment (MichelRottleuthner) [reply tests/xtimer_now_irq/main.c]: Ahh just passing it would feel like lying. Why not just replace the `uint32_t t1 = xtimer_now_usec();` etc. with the 64 bit variants? Then it works for 32 bit too. 
Hmm I guess the problem is then we have no way to CI-run it for 16 bit timers *and* keep it working for manual 32 bit testing?.
How about an additional flag for manual running of 32 bit + some terminal output warning about 32 bit not being checked without that flag?

Comment (kaspar030) [reply tests/xtimer_now_irq/main.c]: > Ahh just passing it would feel like lying. Why not just replace the `uint32_t t1 = xtimer_now_usec();` etc. with the 64 bit variants? Then it works for 32 bit too.

yyyyyes, but isn't that testing something different then?

> How about an additional flag for manual running of 32 bit + some terminal output warning about 32 bit not being checked without that flag?

As long as by default on 32bit on CI, this just prints "not testing anything" "OK", so CI just passes. Could also blacklist.

As is it shouldn't run at all, on 32bit, just so noone actually waits 2**32usec for something to happen. :wink:

File: tests/xtimer_now_irq/README.md (hunk 12 above)
Comment (JulianHolzwarth) [reply tests/xtimer_now_irq/README.md]: What do you want to have there? remove this line? : `(i.e. >72 minutes with a 32 bit timer at 1MHz)`

File: tests/xtimer_now_irq/tests/01-run.py (hunk 13 above)
Comment (JulianHolzwarth) [reply tests/xtimer_now_irq/tests/01-run.py]: yes

**Comment by JulianHolzwarth** [comment]
@MichelRottleuthner @kaspar030 
Should the test only support <24 bits timer?

**Comment by JulianHolzwarth** [comment]
why is this : 

> JulianHolzwarth requested review from fjmolinas, leandrolanzieri, miri64 and smlng as code owners

Because of `TEST_ON_CI_BLACKLIST += all` ?

**Comment by MichelRottleuthner** [comment]
>Should the test only support <24 bits timer?

If anything then <= 24, but actually 32 bit would still be possible when using 64 bit API. In any case 32 Bit must be optional (i.e. only for manual testing). I'd also vote against blacklisting it for all boards - why not CI-run it for 16 and 24 bit?

File: tests/xtimer_now_irq/main.c (hunk 11 above)
Comment (MichelRottleuthner) [reply tests/xtimer_now_irq/main.c]: > yyyyyes, but isn't that testing something different then?

Didn't test yet, but no not really

>As long as by default on 32bit on CI, this just prints "not testing anything" "OK", so CI just passes

I'd prefer some "not testing anything" "OK" over blacklisting.

> just so noone actually waits 2**32usec

I definitely had cases when testing timer stuff that made me desperate enough too run such long running tests^^

**Comment by JulianHolzwarth** [comment]
@MichelRottleuthner 

> If anything then <= 24, but actually 32 bit would still be possible when using 64 bit API. In any case 32 Bit must be optional (i.e. only for manual testing). I'd also vote against blacklisting it for all boards - why not CI-run it for 16 and 24 bit?

I thought @kaspar030 wanted it this way. (only <24 bit and blacklist all)

> I suggest to CI blacklist this test 



> As is it shouldn't run at all, on 32bit, just so noone actually waits 2**32usec for something to happen. wink

**Comment by JulianHolzwarth** [comment]
@MichelRottleuthner @kaspar030 squashed and all comments adressed.

**Comment by MichelRottleuthner** [comment]
ping @kaspar030 all good now?

**Comment by JulianHolzwarth** [comment]
new rebase (no changes)

**Comment by JulianHolzwarth** [comment]
@kaspar030 Can we merge?

**Comment by MichelRottleuthner** [comment]
@kaspar030 would you mind removing your block on this one? All your requests should be addressed.
